#12455 - Fix Missing Cache-Control Headers for Password-Protected Pages

krystianmagdziarz opened the pull request:
Issue #12455

Headers before:

<img width="2048" alt="Zrzut ekranu 2024-10-24 o 17 15 18" src="https://github.com/user-attachments/assets/0bc8971f-fe11-49ed-92be-42a15ee02b56">

Headers after:

<img width="2048" alt="Zrzut ekranu 2024-10-24 o 17 20 13" src="https://github.com/user-attachments/assets/148e2375-7a33-4201-9711-88d832a0861c">

lb- commented:
@krystianmagdziarz thank you for the PR. 

Can you review the CI failure (looks like formatting only). 

In addition, could you please add a basic unit test for each of these views, just so that future changes don't accidentally remove the headers. 

FYI @RealOrangeOne

RealOrangeOne reviewed (commented):
I think this is a great start - thanks for looking into this!

However, you'll also need to add the headers to the response itself, not just the password protection page. That way the page itself won't be cached, bypassing auth.

krystianmagdziarz commented:
Thank you for the review! Here are the updates:

* Added "no-cache" headers for pages after form submission (behind a password);
* Added tests;
* Improved code formatting.

Let me know if there's anything else you'd like me to adjust!

No password page:
<img width="2048" alt="image" src="https://github.com/user-attachments/assets/8846cd0e-5009-4c11-a3e0-2d4aaeab76c5">

Password page:
<img width="2048" alt="image" src="https://github.com/user-attachments/assets/954f3ef4-e0a6-402e-8b37-68d18c17da75">

gasman commented on wagtail/models/__init__.py:
```diff
@@ -2076,12 +2076,21 @@ def get_preview_template(self, request, mode_name):
+            self.get_view_restrictions()
+            .filter(restriction_type=BaseViewRestriction.PASSWORD)
+            .exists()
+        ):
```
This is adding an extra query on every page request, duplicating the one in the `before_serve_page` hook - which is something I'd like to avoid.

The original intent of using the `before_serve_page` hook was to keep all view restriction logic out of this serve method, so that it's performing one focused task - if that's no longer possible (because we need to modify the response from the page, not just block it), then we might as well move the `before_serve_page` code here so that we only have to do the query once.

However, the _secondary_ purpose of the `before_serve_page` hook was as a sort of "dogfooding" exercise - making sure our hooks are powerful enough to implement features that are useful to Wagtail itself, so that we can be confident that they're powerful enough for the things third-party developers might want. For that reason, I'd much prefer to improve our hook capabilities so that we can do everything outside of `serve`. That might mean introducing a new kind of hook that behaves more like Django middleware, so that control returns to the hook after the response has been generated.

krystianmagdziarz replied:
Do you mean to create a new hook after_serve_page and move all logic there?

gasman replied:
That's one possibility, but presumably the permission test will still have to go in `before_serve_page` so we'll still be stuck with running the query twice. I'm thinking a better approach might be more like https://docs.djangoproject.com/en/5.1/topics/http/middleware/#writing-your-own-middleware - where the hook function receives a callable that it can call to continue the request/response cycle (or ignore it and return its own response). That way, the 'before' and 'after' parts can happen in the same function.

krystianmagdziarz replied:
Why can't we move all the logic from `before_serve_page` to `after_serve_page`?  Only instead of returning `None` when there are no restrictions, return the original `HttpResponse` served by the Page `serve` method. Then it doesn't add an extra query.

gasman replied:
That would mean the page's `serve` method gets executed even if the user can't view the page. At the very least, that would be inefficient (since the server will be running queries and doing other processing to generate the page, only for it to be thrown away) - in the worst case, it could have side effects such as handling a form submission, meaning that anonymous users could submit forms that are only supposed to be accessible to privileged users.

krystianmagdziarz replied:
You're right.
I will propose something like this next week.

krystianmagdziarz replied:
Hey,

Together with Dawid Bugajewski (@buugaj), we reviewed the requirements and collaboratively developed a solution. This approach introduces a new hook that functions similarly to Django middleware, as initially suggested.

@gasman Could you please verify if this solution is acceptable? I’d also appreciate your feedback before we proceed with writing unit tests.

`wagtail/views.py`

```python
def serve_chain(page, request, args, kwargs):
    return page.serve(request, *args, **kwargs)

def serve(request, path):
    route_result = Page.route_for_request(request, path)
    if route_result is None:
        raise Http404
    else:
        page, args, kwargs = route_result

    on_serve_chain = serve_chain
    for fn in reversed(hooks.get_hooks("on_serve_page")):
        on_serve_chain = fn(on_serve_chain)
    
    for fn in hooks.get_hooks("before_serve_page"):
        result = fn(page, request, args, kwargs)
        if isinstance(result, HttpResponse):
            return result
        
    return on_serve_chain(page, request, args, kwargs)
```


`wagtail/wagtail_hooks.py`

```python
@hooks.register("on_serve_page")
def check_view_restrictions(callback):
    def inner(page, request, serve_args, serve_kwargs):
        """
        Check whether there are any view restrictions on this page which are
        not fulfilled by the given request object. If there are, return an
        HttpResponse that will notify the user of that restriction (and possibly
        include a password / login form that will allow them to proceed). If
        there are no such restrictions, return None
        """
        restrictions = page.get_view_restrictions()
        response = None
        for restriction in restrictions:
            if not restriction.accept_request(request):
                if restriction.restriction_type == PageViewRestriction.PASSWORD:
                    from wagtail.forms import PasswordViewRestrictionForm

                    form = PasswordViewRestrictionForm(
                        instance=restriction,
                        initial={"return_url": request.get_full_path()},
                    )
                    action_url = reverse(
                        "wagtailcore_authenticate_with_password",
                        args=[restriction.id, page.id],
                    )

                    response = page.serve_password_required_response(
                        request, form, action_url
                    )
                    add_never_cache_headers(response)
                    return response
                elif restriction.restriction_type in [
                    PageViewRestriction.LOGIN,
                    PageViewRestriction.GROUPS,
                ]:
                    response = require_wagtail_login(next=request.get_full_path())
                    add_never_cache_headers(response)
                    return response
                
        response = callback(page, request, serve_args, serve_kwargs)
        if restrictions:
            add_never_cache_headers(response)
        return response
    return inner
```

Comments:

* serve_chain is a base function that calls page.serve(request, *args, **kwargs), responsible for rendering the page.
* The hook on_serve_page is then applied to serve_chain, wrapping it with any additional functions registered under this hook. These functions are processed in reverse order (reversed) to ensure the last-added hook wraps the base serve_chain.
* The resulting on_serve_chain is a chained series of functions, which allows hooks to intercept or modify the page rendering process dynamically.
* Before calling on_serve_chain, any functions registered under before_serve_page are executed. If one of these hooks returns an HttpResponse, it immediately stops further processing, returning the response as the final output (useful for access control, for instance).
* If none of the before_serve_page hooks returns an HttpResponse, the on_serve_chain function is called, completing the request handling process by rendering the page.

gasman replied:
@krystianmagdziarz Sorry for the delay in replying! The `on_serve_page` chaining approach looks perfect to me.

krystianmagdziarz commented:
@gasman what is the status of the issue?

gasman commented:
Thanks for the updates @krystianmagdziarz! Please could you add documentation for the new `on_serve_page` hook in https://github.com/wagtail/wagtail/blob/main/docs/reference/hooks.md?

krystianmagdziarz commented:
@gasman done

gasman commented on docs/reference/hooks.md:
```diff
@@ -1118,6 +1118,42 @@ def block_googlebot(page, request, serve_args, serve_kwargs):
+- Modifying the response content
+- Adding logging or monitoring
+
+Note that if you need to prevent the page from being served entirely, you should use the `before_serve_page` hook instead.
```
I don't think this is necessary - an `on_serve_page` hook can still prevent the page from being served by not calling `next_serve_page`, just like the `check_view_restrictions` one does when serving the 'password required' response.

gasman reviewed (approved):
Thanks for your work on this! Just one nitpick with the documentation, but I can fix that at the point of merging.

gasman commented:
Merged in 32417f9adca1d507bb4ca4880ee4d8879062de04. Thanks @krystianmagdziarz and @buugaj!
